rpc: reduce certificate-related allocations during authentication

mgartner opened the pull request:
#### rpc: add authenication benchmark

Release note: None

#### rpc: reduce certificate-related allocations during authentication

This commit avoids certificate-related allocations in the happy path of
authenticating network requests.

Release note: None

#### security: inline getCertificatePrincipals

The `getCertificatePrincipals` function has been inlined into its one
callsite to avoid allocating a slice of the results.

Fixes #133317

Release note: None

mgartner committed:
rpc: reduce certificate-related allocations during authentication

This commit avoids certificate-related allocations in the happy path of
authenticating network requests.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/134228)
<!-- Reviewable:end -->

mgartner committed:
security: inline getCertificatePrincipals

The `getCertificatePrincipals` function has been inlined into its one
callsite to avoid allocating a slice of the results.

Release note: None

mgartner committed:
rpc: add authentication benchmark

Release note: None

mgartner commented:
This contributed to 1% of all allocations in a sysbench workload. I was unable to measure an improvement in the sysbench benchmarks added in #133132. But the microbenchmark I added shows the improvement:

```
name                     old time/op    new time/op    delta
Authenticate/tenTen        7.48µs ± 1%    7.65µs ± 1%     ~     (p=0.057 n=4+4)
Authenticate/rootDN        9.72µs ± 2%    9.60µs ± 1%     ~     (p=0.548 n=5+5)
Authenticate/nodeDN        10.2µs ±10%     9.6µs ± 2%     ~     (p=0.310 n=5+5)
Authenticate/commonRoot    7.62µs ± 1%    7.62µs ± 1%     ~     (p=0.841 n=5+5)
Authenticate/commonNode    7.55µs ± 1%    7.70µs ± 1%   +1.94%  (p=0.008 n=5+5)

name                     old alloc/op   new alloc/op   delta
Authenticate/tenTen        57.4kB ± 0%    57.4kB ± 0%     ~     (all equal)
Authenticate/rootDN        57.9kB ± 0%    57.9kB ± 0%     ~     (all equal)
Authenticate/nodeDN        57.9kB ± 0%    57.9kB ± 0%     ~     (all equal)
Authenticate/commonRoot    57.5kB ± 0%    57.4kB ± 0%   -0.11%  (p=0.008 n=5+5)
Authenticate/commonNode    57.5kB ± 0%    57.4kB ± 0%   -0.11%  (p=0.008 n=5+5)

name                     old allocs/op  new allocs/op  delta
Authenticate/tenTen          3.00 ± 0%      3.00 ± 0%     ~     (all equal)
Authenticate/rootDN          28.0 ± 0%      28.0 ± 0%     ~     (all equal)
Authenticate/nodeDN          28.0 ± 0%      28.0 ± 0%     ~     (all equal)
Authenticate/commonRoot      5.00 ± 0%      3.00 ± 0%  -40.00%  (p=0.008 n=5+5)
Authenticate/commonNode      5.00 ± 0%      3.00 ± 0%  -40.00%  (p=0.008 n=5+5)
```

stevendanna reviewed (commented):
Nice! 

> Reviewed 1 of 1 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134228#-:-OAvoLyI7CeAJT8vR6ST:bwuv9up)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @nvanbenschoten, and @tbg)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134228#-OAvgCpgESfazCPWbZ6d):*
[nit] s/authenication/authentication/
___
*[`pkg/rpc/auth_test.go` line 292 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134228#-OAvhfen67tCRSvNGSmw:-OAvhfen67tCRSvNGSmx:b-auer19) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d1bfdd89f49c0ae41bfa902fd9ff3c8e0f7fc75/pkg/rpc/auth_test.go#L292)):*
> ```Go
> 		{name: "nodeDN", systemID: stid, ous: nil, commonName: "foo", nodeDNString: "CN=foo"},
> 		{name: "commonRoot", systemID: stid, ous: nil, commonName: "root"},
> 		{name: "commonNode", systemID: stid, ous: nil, commonName: "node"},
> ```

Should we add a case with a SubjectAltName? I imagine we don't see these often in practice so perhaps it isn't worth it.


<!-- Sent from Reviewable.io -->

tbg commented:
Do you want this PR to at least mention (but probably fix) https://github.com/cockroachdb/cockroach/issues/133317?
The microbenchmark simply doesn't seem to execute this code path (https://github.com/cockroachdb/cockroach/issues/133317#issuecomment-2457304945), otherwise it would've likely shown up.

tbg commented on pkg/security/auth.go:
```diff
@@ -205,42 +205,77 @@ func getCertificatePrincipals(cert *x509.Certificate) []string {
+func CertificateUserScopeContainsFunc(
+	peerCert *x509.Certificate, fn func(CertificateUserScope) bool,
+) (ok bool, _ error) {
+	res := false
```
this escapes to the heap (?) but could be sync.Pooled in principle.

mgartner replied:
This doesn't escape.

tbg commented on pkg/rpc/auth_test.go:
```diff
@@ -274,6 +274,72 @@ func TestAuthenticateTenant(t *testing.T) {
+		rootDNString string
+		nodeDNString string
+	}{
+		{name: "tenTen", systemID: stid, ous: correctOU, commonName: "10"},
```
Which of the benchmarks reflects the case we're seeing in the sysbench reference workload? Could you add inline comments about when we expect to hit which?

tbg commented on pkg/rpc/auth_test.go:
```diff
@@ -274,6 +274,72 @@ func TestAuthenticateTenant(t *testing.T) {
+			b.ResetTimer()
+			var clusterSettings map[settings.InternalKey]settings.EncodedValue
+			for i := 0; i < b.N; i++ {
+				_, err := rpc.TestingAuthenticateTenant(ctx, tc.systemID, clusterSettings)
```
I checked out your PR to look at the flamegraph for the remaining three allocs -  `TestingAuthenticateTenant` is not a good entry point for this benchmark as it makes a `settings.NewUpdater` etc, which dominates the allocations:


```
./dev bench --ignore-cache --stream-output --bench-mem ./pkg/rpc --filter Authenticate/commonRoot --test-args '-test.memprofile=mem.pb.gz -test.benchtime=50000x -test.cpu=8' && pprof -http :6060 mem.pb.gz
```

<img width="1728" alt="image" src="https://github.com/user-attachments/assets/aedb335b-f426-460c-b9bd-e1019ad3002e">

Could you refactor this entry point to take a `settings.Values` instead, or make a new one?

I hacked this using this diff

and the result shows that we're actually allocation-free on the `commonRoot` and `commonName` benchmarks.

```
BenchmarkAuthenticate/tenTen-8             50000                43.08 ns/op            0 B/op          0 allocs/op
BenchmarkAuthenticate/rootDN-8             50000              1165 ns/op             464 B/op         25 allocs/op
BenchmarkAuthenticate/nodeDN-8             50000              1149 ns/op             464 B/op         25 allocs/op
BenchmarkAuthenticate/commonRoot-8         50000                80.63 ns/op            0 B/op          0 allocs/op
BenchmarkAuthenticate/commonNode-8         50000                80.48 ns/op            0 B/op          0 allocs/op
```

tbg replied:
For completeness here's the alloc_objects for the combined `BenchmarkAuthenticate` with the spurious allocs suppressed
<img width="1722" alt="image" src="https://github.com/user-attachments/assets/9a065abb-2963-490a-a536-0501a4ae1141">

mgartner replied:
Good idea. I was lazy and just stopped tinkering with the benchmark when I observed the reduction in allocs. The benchmark is more accurate now.

I'll look into the `CheckCertDNMatchesRootDNorNodeDN` allocations and see if they can be reduced easily. I have zero experience here, so it'll take me some time. I also don't want to spend too many cycles on it since it's not coming up in our target workload.

tbg replied:
Yep, if it's not in the target let's not spend more time on it now.
<!-- Reviewable comment -OBziVruDvxhf9p9gORt:bri31if -->
<!-- Sent from Reviewable.io -->

tbg reviewed (changes requested):
Thanks for tackling this Marcus! Haven't fully reviewed yet, just a comment on the benchmark in the first commit. Will give a full review when that's addressed.

mgartner commented:
>Haven't fully reviewed yet, just a comment on the benchmark in the first commit. Will give a full review when that's addressed.

The changes are pretty simple if you hide the whitespace changes.

mgartner commented:
The new benchmark before/after:

```
name                     old time/op    new time/op    delta
Authenticate/tenTen        42.5ns ± 5%    42.2ns ± 2%      ~     (p=0.794 n=5+5)
Authenticate/rootDN        1.23µs ± 1%    1.21µs ± 0%    -1.85%  (p=0.008 n=5+5)
Authenticate/nodeDN        1.23µs ± 1%    1.21µs ± 0%    -1.24%  (p=0.008 n=5+5)
Authenticate/commonRoot     127ns ± 3%      76ns ± 0%   -39.82%  (p=0.008 n=5+5)
Authenticate/commonNode     129ns ± 4%      76ns ± 0%   -40.70%  (p=0.008 n=5+5)

name                     old alloc/op   new alloc/op   delta
Authenticate/tenTen         0.00B          0.00B           ~     (all equal)
Authenticate/rootDN          464B ± 0%      464B ± 0%      ~     (all equal)
Authenticate/nodeDN          464B ± 0%      464B ± 0%      ~     (all equal)
Authenticate/commonRoot     64.0B ± 0%      0.0B       -100.00%  (p=0.008 n=5+5)
Authenticate/commonNode     64.0B ± 0%      0.0B       -100.00%  (p=0.008 n=5+5)

name                     old allocs/op  new allocs/op  delta
Authenticate/tenTen          0.00           0.00           ~     (all equal)
Authenticate/rootDN          25.0 ± 0%      25.0 ± 0%      ~     (all equal)
Authenticate/nodeDN          25.0 ± 0%      25.0 ± 0%      ~     (all equal)
Authenticate/commonRoot      2.00 ± 0%      0.00       -100.00%  (p=0.008 n=5+5)
Authenticate/commonNode      2.00 ± 0%      0.00       -100.00%  (p=0.008 n=5+5)
```

mgartner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134228)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @nvanbenschoten, @stevendanna, and @tbg)

___
*[`-- commits` line 2 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134228#-OAvgCpgESfazCPWbZ6d:-OBzfK1JCwylYcJ_rK2c:b-896fix):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

[nit] s/authenication/authentication/
</blockquote></details>

Done.
___
*[`pkg/rpc/auth_test.go` line 292 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/134228#-OAvhfen67tCRSvNGSmw:-OBzhc28DcMEFzc7Hoys:b-jboiqi) ([raw file](https://github.com/cockroachdb/cockroach/blob/6d1bfdd89f49c0ae41bfa902fd9ff3c8e0f7fc75/pkg/rpc/auth_test.go#L292)):*
<details><summary><i>Previously, stevendanna (Steven Danna) wrote…</i></summary><blockquote>

Should we add a case with a SubjectAltName? I imagine we don't see these often in practice so perhaps it isn't worth it.
</blockquote></details>

I'm not familiar enough with this code to know whether it'd be useful or not. I added seemingly basic cases and two cases that hit the code path I optimized. Happy to add more if you'd like.


<!-- Sent from Reviewable.io -->

tbg reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/134228#-:-OBzia4f4ebiIpDHgLOx:bnfp4nl)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @nvanbenschoten, and @stevendanna)



<!-- Sent from Reviewable.io -->

mgartner commented:
TFTRs!

bors r+
